File: epidata_test_fe/src/domain/use-cases/TaskUseCase.ts
import { Task, UpdateTaskRequest } from "../entities/Task";
import { TaskRepository } from "../repositories/TaskRepository";

export class TaskUseCase {
  constructor(private repo: TaskRepository) {}

  async getAllTasks(): Promise<Task[]> {
    return this.repo.getAll();
  }

  async getTaskById(taskId: number): Promise<Task> {
    return this.repo.getById(taskId);
  }

  async createTask(title: string): Promise<Task> {
    if (!title.trim()) {
      throw new Error("Task title cannot be empty");
    }
    return this.repo.create(title);
  }

  async updateTask(taskId: number, task: UpdateTaskRequest): Promise<Task> {
    if (!task.title.trim()) {
      throw new Error("Task title cannot be empty");
    }
    return this.repo.update(taskId, task);
  }

  async markTaskComplete(taskId: number): Promise<Task> {
    return this.repo.markComplete(taskId);
  }

  async deleteTask(taskId: number): Promise<void> {
    return this.repo.delete(taskId);
  }
}

File: epidata_test_be/app/use_cases/task_use_case.py
from app.domain.models.task_entity import Task
from app.domain.repositories.task_repository import TaskRepository

class TaskUseCase:
    def __init__(self, repo: TaskRepository):
        self.repo = repo
        
    def list_tasks(self):
        return self.repo.get_all()
    
    def get_task_by_id(self, task_id: int):
        return self.repo.get_by_id(task_id)
    
    def create_task(self, title: str):
        return self.repo.create(title)
    
    def update_task(self, task_id: int, task: Task):
        return self.repo.update(task_id, task)
    
    def mark_task_complete(self, task_id: int):
        return self.repo.mark_complete(task_id)
    
    def delete_task(self, task_id: int):    
        return self.repo.delete(task_id)